kythe/typescript/README.md

```diff
@@ -1,5 +1,269 @@
 # Kythe indexer for TypeScript
 
+## VName Specification
+
+This specification defines how the indexer expresses VNames for TypeScript
+declarations. You may find this useful if you are developing an application that
+relies on TypeScript code you don't want to re-index.
+
+A formal listing of the specification is provided below, but the
+[spec tests](./testdata/declaration_spec.ts) may be more useful for real
+examples.
+
+### VName signature
+
+The signature description language used in this document is similar to regex.
+
+-   A substring in `\b\$.*\b` is a variable.
+    -   e.g. `$DECLARATION_NAME` is a variable refering to the name of
+        declaration.
+-   A substring in `\[.*\]\?` is matched 0 or 1 times.
+    -   e.g. `(hidden)?` is really `hidden` or ``.
+-   A substring in `\[.*\]\*` is matched 0 or more times.
+    -   e.g. `[a]*` is ``, or`a`, or`aa`, ...
+-   All other substrings are literals.
+    -   e.g. `get#$NAME` is really `get#foo` if `$NAME = foo`.
+
+The signature of a TypeScript declaration is defined by the following schema:
+
+```regex
+$PART[.$PART]*[#type]?
+```
+
+where `$PART` is a component of the enclosing declaration scope and `#type` is
+appended to the signature of type declarations. `SyntaxKind`s that are type
+declarations are:
+
+-   `ClassDeclaration` (also a value)
+-   `EnumDeclaration` (also a value)
+-   `InterfaceDeclaration`
+-   `TypeAliasDeclaration`
+-   `TypeParameter`
+
+As an example of this schema, in
+
+```typescript
+class A {
+  public foo: string;
+}
+
+type B = A;
+```
+
+`foo` has the signature `A.foo` and `B` has the signature `B#type`.
+
+Signature components (`$PART`s in the schema) are defined below.
+
+#### File Module
```

Comment:
To keep this spec machine-checked, I wonder if it's better to stuff all of this text as comments within the test.

Reply:
I worry that will make reading the spec much harder, though. The schema shouldn't have many breaking changes from when it's merged, so I don't anticipate this file having to be updated too often. We link to the test file at the top of the schema; if only GitHub supported embedded code snippets in Markdown...